### DISCUSSION TITLE
Throw custom error per rule

### DISCUSSION THREAD

**Discussion by MahaZaid** [taocpp/PEGTL#203] [Q&A] [answered]
I have a grammar as follows 
```
struct andOperator : TAO_PEGTL_STRING("AND") {};
struct keyword: minus<identifier, andOperator >{};
struct atomic: seq< keyword, andOperator , keyword >{};
struct expression: must<atomic> {};
```

And I have a custom error message for every rule as follows:
```
template <>inline const std::string errors<query:: andOperator >::error_message = "andOperator";
template <>inline const std::string errors<query:: keyword >::error_message = "invalid keyword";
template <>inline const std::string errors<query:: atomic >::error_message = "atomic";
template <>inline const std::string errors<query:: expression >::error_message = "expression";
```

If my query is for example `AND` so it will not match the expression but it will fail in the rule `atomic`, so the error message I get is `atomic`.

my question: is there any way to track the problem back to `keyword` rule 
so that the error message that will be thrown would be `invalid keyword`, the one specified in `keyword` error message ?

**Comment by d-frey** [comment, marked as answer]
You could unroll `minus` back to `rematch` and use the inner rule to raise an error directly:

```
struct andOperator : TAO_PEGTL_STRING("AND") {};
struct invalidKeyword {};
struct keyword: rematch<identifier, not_at< andOperator, eof, raise< invalidKeyword > > > {};
struct atomic: seq< keyword, plus<space>, andOperator, plus<space>, keyword >{};
struct expression: must<atomic> {};
```

Now it will use the error message specialization for `invalidKeyword` when an identifier matches a keyword.

**Comment by ColinH** [comment]
I'm assuming that your question was answered, if not please feel free to continue here and/or reopen this issue.
